build-system/eslint-rules/await-expect.js

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright 2019 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+'use strict';
+
+module.exports = function(context) {
+  return {
+    CallExpression(node) {
+      const filename = context.getFilename();
+      if (!/test-e2e|\/test\/e2e\//.test(filename)) {
+        return;
+      }
+
+      const {callee} = node;
+      if (callee.type !== 'Identifier') {
+        return;
+      }
+
+      if (callee.name !== 'expect') {
+        return;
+      }
+
+      if (hasAwaitParent(node)) {
```

Comment:
`expect()` returns a `Promise`? If so, then the direct parent should be an await.

Reply:
I looked through the AST and if the `expect` is followed by properties like `expect(blah).to.be.truthy` the parent of the `expect` `CallExpression` node is a chain of `MemberExpression`s.

```js
const espree = require('espree');
const code = 'async function foo() { await expect("foo").to.be.truthy; }';
const programNode = espree.parse(code, {ecmaVersion: 2017});
const fnBodyNode = programNode.body[0].body.body[0];
const awaitExpression = fnBodyNode.expression;
// Node {
//   type: 'AwaitExpression',
//   start: 23,
//   end: 55,
//   argument:
//    Node {
//      type: 'MemberExpression',
//      start: 29,
//      end: 55,
//      object:
//       Node {
//         type: 'MemberExpression',
//         start: 29,
//         end: 48,
//         object: [Node],
//         property: [Node],
//         computed: false },
//      property:
//       Node { type: 'Identifier', start: 49, end: 55, name: 'truthy' },
//      computed: false } }
```

Reply:
What value are we waiting on, then?

Reply:
oh right, I should give a better example.

```
await expect(controller.getElementAttribute(elementHandle, 'src')).to.equal('foo');
```

`controller.getElementAttribute` returns a Promise, that is capable of being re-queried until it matches the expected value (`ControllerPromise` in `functional-test-controller.js`](https://github.com/ampproject/amphtml/blob/680dd188e47114af397083500333a7ab0ddd886e/build-system/tasks/e2e/functional-test-controller.js#L57)). `expect` resolves after a timeout or after the expectation matches.